uast/ann/ann.go

```diff
@@ -14,9 +17,43 @@ const (
 	descendantOrSelf
 )
 
-// Predicate is a function that takes a node and returns a boolean. It
-// is provided as a convenient way of defining a PathMatcher.
-type Predicate func(n *uast.Node) bool
+// String returns the XPath (XML Path Language) string representation of
+// an axis.
+func (a axis) String() string {
+	switch a {
+	case self:
+		return "self"
+	case child:
+		return "child"
+	case descendant:
+		return "descendant"
+	case descendantOrSelf:
+		return "descendant-or-self"
+	default:
+		panic(fmt.Sprintf("unknown axis: %q", a))
+	}
+}
+
+// Predicate is the interface that wraps boolean tests for uast.Nodes.
+//
+// The Eval function evaluates the test over a node and returns its
+// boolean output.  The String function returns  a description of the
+// predicate in a syntax similar to XPath.
+type Predicate interface {
+	fmt.Stringer
+	Eval(n *uast.Node) bool
+}
+
+// predicate implements Predicate by using desc as the XPath
+// representation of the predicate.
+type predicate struct {
```

Comment:
I wonder if the XPath representation could be decoupled from the predicate and action interfaces/implementation. It looks like this repr is ortoghonal to both predicates and actions (i. e. we could have n different representations and they shouldn't affect how actions and predicates are implemented.)